### PR TITLE
Implement UserState

### PR DESCRIPTION
**Pull request by gpang**
The main contribution is to remove the LoginUser singleton, and introduce a new concept called `UserState`. The `UserState` manages the login information and state of any client (master, worker, client). There are specific implementations for UserState depending on the security type, so every implementation has its own class.

### CODE REVIEW

**Comment by gpang** [comment]
@bf8086 @ZacBlanco

File: core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java [hunk 1]
```diff
@@ -102,16 +116,21 @@ public OutStreamOptions(CreateFilePOptions options, AlluxioConfiguration alluxio
...
         alluxioConf);
     mWriteTier = alluxioConf.getInt(PropertyKey.USER_FILE_WRITE_TIER_DEFAULT);
     mWriteType = alluxioConf.getEnum(PropertyKey.USER_FILE_WRITE_TYPE_DEFAULT, WriteType.class);
-    mOwner = SecurityUtils.getOwnerFromLoginModule(alluxioConf);
-    mGroup = SecurityUtils.getGroupFromLoginModule(alluxioConf);
+    try {
+      mOwner = context.getUserState().getUser().getName();
+      mGroup = CommonUtils.getPrimaryGroupName(mOwner, context.getConf());
+    } catch (IOException e) {
+      mOwner = "";
```
Comment (bf8086) [inline core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java]: (nit) Log an error message for the exception

File: core/common/src/main/java/alluxio/security/user/TestUserState.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.exception.status.UnauthenticatedException;
+import alluxio.security.User;
+
+import javax.security.auth.Subject;
+
+/**
+ * UserState for testing only.
+ */
+public class TestUserState extends BaseUserState {
```
Comment (bf8086) [inline core/common/src/main/java/alluxio/security/user/TestUserState.java]: Can we move this under the test folder?

File: tests/src/test/java/alluxio/client/cli/fs/command/SetFaclCommandIntegrationTest.java [hunk 3]
```diff
@@ -12,12 +12,16 @@
...
 
 import alluxio.AlluxioURI;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.FileSystemContext;
 import alluxio.conf.PropertyKey;
 import alluxio.client.cli.fs.AbstractFileSystemShellTest;
 import alluxio.client.file.FileSystemTestUtils;
 import alluxio.client.file.URIStatus;
+import alluxio.conf.ServerConfiguration;
 import alluxio.exception.AlluxioException;
```
Comment (bf8086) [inline tests/src/test/java/alluxio/client/cli/fs/command/SetFaclCommandIntegrationTest.java]: (nit) This is not introduced by your change, can you reorder the import?

File: core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java (hunk 1 above)
Comment (ZacBlanco) [reply core/client/fs/src/main/java/alluxio/client/file/options/OutStreamOptions.java]: We could set the level to `DEBUG` if we ever need to trace security exceptions?

File: core/common/src/main/java/alluxio/security/CurrentUser.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(mName, mServiceName);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    return Objects.equal(mName, mServiceName);
```
Comment (ZacBlanco) [inline core/common/src/main/java/alluxio/security/CurrentUser.java]: Why is equals testing whether `mName == mServiceName`? Shouldn't it be something like:
```java
    public boolean equals(Object o) {
      if (!(o instanceof CurrentUser)) {
        return false;
      }
      CurrentUser other = (CurrentUser) o;
      return Objects.equals(mName, other.mName)
          && Objects.equals(mServiceName, other.mServiceName);
    }
```

File: core/common/src/main/java/alluxio/security/user/BaseUserState.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (this.getClass() != o.getClass()) {
+      return false;
+    }
+    BaseUserState that = (BaseUserState) o;
+    return Objects.equal(mSubject, that.mSubject);
```
Comment (ZacBlanco) [inline core/common/src/main/java/alluxio/security/user/BaseUserState.java]: should we also compare `mUser`? Same for hashCode

File: core/common/src/main/java/alluxio/security/CurrentUser.java (hunk 4 above)
Comment (gpang) [reply core/common/src/main/java/alluxio/security/CurrentUser.java]: good catch! fixed

File: core/common/src/main/java/alluxio/security/user/BaseUserState.java (hunk 5 above)
Comment (gpang) [reply core/common/src/main/java/alluxio/security/user/BaseUserState.java]: Subject equality will compare the principals as well. `mUser` is basically a cached reference of the User in the Subject.

**Comment by gpang** [comment]
alluxio-bot, merge this please

**Pull request merged**